src/widgets/query-rule-custom-data/__tests__/query-rule-custom-data-test.ts

```diff
@@ -59,7 +64,11 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/query-rule-
           container: document.createElement('div'),
         });
 
-        widget.init({ helper, state: helper.state, instantSearchInstance: {} });
+        widget.init!({
```

Comment:
Since there's no guarantee that the `init` or `render` method of the widget exists, we need to force them with `!` in the tests.

Reply:
It is not possible to solve this using [conditional types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#conditional-types) to get the correct return type from `queryRuleCustomData`? (my fear is that the users get the same issue when calling `widget.init`)

Reply:
We tried some stuff like conditional types with @samouss but each solution brought other complications. @samouss has more input than me on this.

I don't think users have to manipulate `widget.init` except in tests as we do.

Another solution would be to enforce `init`, `render` and `dispose` from now on. WDYT?

Reply:
Well, I still lack context on the lib to see the impact on enforcing these methods.

But I would be very interested to see what complications you encountered with @samouss with conditional types (for my own curiosity :wink: )

Reply:
> But I would be very interested to see what complications you encountered with @samouss with conditional types (for my own curiosity 😉 )

We've tried with both union types and conditional types. With the latter we're almost there but not yet. Once we put the widgets inside an array we're not able to find which kind of widget it is. It means that we only have access to all the common attributes of each kind of widget (which makes perfect sense). Here is [an example](https://www.typescriptlang.org/play/index.html#src=type%20InitOptions%20%3D%20any%3B%0D%0Atype%20RenderOptions%20%3D%20any%3B%0D%0Atype%20DisposeOptions%20%3D%20any%3B%0D%0Atype%20SearchParameters%20%3D%20any%3B%0D%0A%0D%0Aexport%20interface%20BaseWidget%20%7B%0D%0A%20%20dispose%3F()%3A%20SearchParameters%3B%0D%0A%20%20getConfiguration%3F()%3A%20SearchParameters%3B%0D%0A%7D%0D%0A%0D%0Aexport%20interface%20WidgetWithInit%20extends%20BaseWidget%20%7B%0D%0A%20%20init(options%3A%20InitOptions)%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Aexport%20interface%20WidgetWithRender%20extends%20BaseWidget%20%7B%0D%0A%20%20render(options%3A%20RenderOptions)%3A%20void%3B%0D%0A%7D%0D%0A%0D%0Aexport%20interface%20WidgetWithInitRender%20extends%20WidgetWithInit%2C%20WidgetWithRender%20%7B%0D%0A%0D%0A%7D%0D%0A%0D%0Aexport%20type%20Widget%20%3D%0D%0A%20%20%7C%20WidgetWithInit%0D%0A%20%20%7C%20WidgetWithRender%0D%0A%20%20%7C%20WidgetWithInitRender%0D%0A%0D%0Aexport%20type%20CreateWidget%3CTWidget%20extends%20Widget%20%3D%20Widget%3E%20%3D%20()%20%3D%3E%20%0D%0A%20%20TWidget%20extends%20WidgetWithInitRender%20%3F%20WidgetWithInitRender%20%3A%0D%0A%20%20TWidget%20extends%20WidgetWithInit%20%3F%20WidgetWithInit%20%3A%0D%0A%20%20WidgetWithRender%0D%0A%0D%0Aconst%20createWidget%3A%20CreateWidget%3CWidgetWithInitRender%3E%20%3D%20()%20%3D%3E%20(%7B%0D%0A%20%20init()%20%7B%7D%2C%0D%0A%20%20render()%20%7B%7D%2C%0D%0A%7D)%3B%0D%0A%0D%0Aconst%20createWidgetInit%3A%20CreateWidget%3CWidgetWithInit%3E%20%3D%20()%20%3D%3E%20(%7B%0D%0A%20%20init()%20%7B%7D%2C%0D%0A%7D)%3B%0D%0A%0D%0Aconst%20createWidgetRender%3A%20CreateWidget%3CWidgetWithRender%3E%20%3D%20()%20%3D%3E%20(%7B%0D%0A%20%20render()%20%7B%7D%2C%0D%0A%7D)%3B%0D%0A%0D%0Aconst%20w%20%3D%20createWidget()%0D%0Aconst%20wInit%20%3D%20createWidgetInit()%3B%0D%0Aconst%20wRender%20%3D%20createWidgetRender()%3B%0D%0Aconst%20ws%20%3D%20%5Bw%2C%20wInit%2C%20wRender%5D%3B%0D%0A%0D%0Aw.init(%7B%7D)%0D%0Aw.render(%7B%7D)%0D%0A%0D%0AwInit.init(%7B%7D)%0D%0AwInit.render(%7B%7D)%0D%0A%0D%0AwRender.init(%7B%7D)%0D%0AwRender.render(%7B%7D)%0D%0A%0D%0Aws.forEach(w%20%3D%3E%20%7B%0D%0A%20%20if%20(typeof%20w.init%20!%3D%3D%20'function')%20%7B%0D%0A%20%20%20%20w.init()%3B%0D%0A%20%20%7D%0D%0A%7D)). Note that I might be completely wrong about the subject though.

> my fear is that the user get the same issue when calling `widget.init`

This is not an issue because we call the methods for the users. They never have to call it directly.

> Another solution would be to enforce init, render and dispose from now on. WDYT?

Yes that would solve the issue but I would avoid to do it right now. I would rather suggest to take this into consideration once we revamp the core of InstantSearch. Until now we can treat them as optional inside the codebase it was (almost) already the case.

https://github.com/algolia/instantsearch.js/blob/ba6e2e61c16c1b524d24cf72a66cbb354cf81b55/src/lib/InstantSearch.js#L387

https://github.com/algolia/instantsearch.js/blob/ba6e2e61c16c1b524d24cf72a66cbb354cf81b55/src/lib/InstantSearch.js#L413